### DISCUSSION TITLE
GSoC 2024: WAL Infrastructure

### DISCUSSION THREAD

**Discussion by sh-soltanpour** [pgmoneta/pgmoneta#288] [Ideas]
## Intro
Hi everyone,

My name is Shahryar (pronounced as shah-ree-yar), and I'm selected as a contributor for `GSoC 2024` in the `PostgreSQL` organization and `Pgmoneta - WAL Infrastructure` project. 

In this project, we want to implement an infrastructure in Pgmoneta that can read and parse the WAL files into WAL records. When our infrastructure has a deep understanding of the WAL files and records, we can apply the appropriate WAL records on a base backup that we have, and we can reach the desired point in the restored database, which is a very helpful feature to achieve point-in-time-recovery (PiTR). As a bonus goal, we can even extend this understanding from a physical to a logical understanding. 

By the end of GSoC 2024, I’m aiming to achieve these items: 
Milestone 1: Implement an in-house infrastructure for parsing WAL segment files to WAL records.
Milestone 2: Being able to restore the database using a base backup and applying a series of WAL records to get to the required point. 
(Bonus) Milestone 3: Extract logical information from WAL records and implement a policy definition framework so developers can define some custom logic for their backup database and filter some tables, rows, etc.  

For more details, feel free to check out my initial proposal for this project [here](https://docs.google.com/document/d/194bAxMbiOvQhL4rCH30mlIhHZW9UccbQa2F8EEMCIwI/edit?usp=sharing).

I would be more than happy to receive any suggestions or ideas here or by sending them to my email: shahryar.soltanpour@gmail.com

I will update the sections below during my progress in this project. 

## List of the work done

The main branch for the writing feature is [`feature-wal-write`](https://github.com/sh-soltanpour/pgmoneta/tree/feature-wal-write).

My main work branch for the whole WAL parsing functionality is [`parse-wal-records`](https://github.com/sh-soltanpour/pgmoneta/tree/parse-wal-records).

- Implemented #299 and merged it to the [`parse-wal-records`](https://github.com/sh-soltanpour/pgmoneta/tree/parse-wal-records) branch. 
- Implemented #296 (Including #299, #300 and #301) and made [the PR](https://github.com/pgmoneta/pgmoneta/pull/355)
- The mentioned major PR aims to finish the project's first milestone. It includes implementing the `rm_desc` function for all resource managers, testing across PostgreSQL versions 13-17, and fixing critical bugs. I'm working on fixing the comments and requested changes during the code review. This is a critical step in making the code able to understand the structure of WAL files, after this PR, we can start implementing the `write` function to write the WAL file back to another instance.
- Implemented and merged #355 into main branch. 
- Implemented #372 with #381, with this PR we have a high-level API of working with the WAL file defined in `walfile.h` and the writing functionality. The writing functionality is tested by feeding WAL files generated by Postgres 13 to 17 to Pgmoneta and comparing the output binary file with the source file using the cmp command.
- Implemented #399 to add `pgmoneta-walinfo` CLI tool. It supports WAL file decompression, JSON and raw output support, Display with color palette, Documentation updates for the new CLI. 

## Progress Log

**27/Oct/2024**
This week I've got #399 merged, this means that now we have `pgmoneta-walinfo` tool added to Pgmoneta, it supports
- WAL file decompression
- JSON and raw output support
- Display with color palette
- Documentation updates for the new CLI
.

**20/Oct/2024**
I'm still working on PR #399, to you can follow the progress on the discussions on the PR page.

**13/Oct/2024**

This week, I've been working on `pgmoneta-walinfo`, a CLI tool that will allow users to read and display WAL files. I've already created a WIP PR #399, and now it supports JSON and raw text files. This PR should be ready to go in a few days and then I can probably work on debugging functionalities. 

**06/Oct/2024**
This week I've been able to debug the `pgmoneta_wal_encode_xlog_record` and fix the writing functionality. 
I've finalized PR #381 and it's merged now. with this PR we have a high-level API of working with the WAL file defined in `walfile.h` and the writing functionality. The writing functionality is tested by feeding WAL files generated by Postgres 13 to 17 to Pgmoneta and comparing the output binary file with the source file using the cmp command.
Next week I will probably start working on utility functions with OIDs (I need to ask @jesperpedersen about what we want here), and also debugging functionality. 

**29/Sep/2024**
This week I've implemented the basic code to be able to write the WAL (including its header and data) to a file.
Currently, I'm facing a problem when comparing the generated file with the actual WAL file, until the index 2112, they're completely the same, but after that they start to change, right now I'm debugging to see why they're getting changed, I guess the problem is somewhere in `pgmoneta_wal_encode_xlog_record` function. 

**22/Sep/2024**
This week I've started working on writing the WAL functionality. 
I've implemented reading and writing each page header functionality in the `pgmoneta_write_wal_file` function.

Currently, I'm working on the ability to convert the main data and `bkpblocks` into `char*`  in the `pgmoneta_wal_encode_xlog_record` function.
I've created #381 draft PR so you can track the changes here. [`feature-wal-write`](https://github.com/sh-soltanpour/pgmoneta/tree/feature-wal-write) should be my active working branch for this feature. 

**15/Sep/2024**
- Merged #355 -- So now we have WAL parsing functionality merged and in-place for Pgmoneta.
- Started working on designing high-level APIs, including ` int read_wal_file(char* path, wal_file** wf), int write_wal_file(wal_file* wf, char* path)`

**01/Sep/2024**
Still working on fixing the comments to [this PR](https://github.com/pgmoneta/pgmoneta/pull/355).
Currently going through [this list](https://elastic-traffic-761.notion.site/Postgres-WAL-diff-5d7f873c9d204ff79cb6779262c76326?pvs=74) and applying them one by one. I've already applied 17 -> 16 changes and currently working on 16 -> 15 changes.

**18/Aug/2024**
Working on the comments on [this PR](https://github.com/pgmoneta/pgmoneta/pull/355).

**11/Aug/2024**
This week, I have drafted the PR that adds the ability to parse wal records and print the description of each record (Find the [PR here](https://github.com/pgmoneta/pgmoneta/pull/355)). This is a fairly big PR as all of the resource managers and functionalities are going to be merged at once. Write now I'm working on adding documentation of how the main function works to the code and also fixed some style-related comments and will be working on fixing the rest of them. 

**4/Aug/2024**
This week I cleaned up some parts of the code, for example getting the Postgres version from pgmoneta and had a bit of difficulty testing it because until now I have been testing the project without a server and just by running a custom command.
A couple of minor changes are still remaining and I think I will be able to create the PR by Friday.

**28/July/2024**

This week, I was able to test the code against WAL files generated from versions 13 to 17, I just noticed a subtle difference in how the `BKPIMAGE_COMPRESSED` macro is working starting version 15 and the parser was working okay after fixing this behavior based on the Postgresql version. 
So I think, if we don't want to replace all predefined values that should be extracted from `pg_control` (or maybe other files), I think I can start the initial PR during this week and go through the feedback iteration. 

**21/July/2024**
This week, I spent most of my time fixing the code issues and memory leaks in a Fedora image. At first, I was facing some weird memory allocation errors that I wasn't getting on macOS, but using Valgrind, I could fix them and now it works fine on Fedora with a clean Valgrind report.

Regarding what's remaining, I can say that there are some defined values in my code that should be parsed from the `pg_control` file I guess, and I've used constant values for them at the moment, and also the code only supports Postgres 14 right now and the compatibility should get extended to the 13-17 versions. With these, do you think that we can start having a PR and reviewing it or do you prefer these to be added and then check them all at once? @jesperpedersen 

**14/July/2024**

1- Moved all of the `.h` and `.c` files to a separate folder. 
2- Fixed the naming of Structs to be consistent with the rest of the code. 
3- Fixed struct definition styles.
4- Fixed some memory allocation bugs happening undercover.
5- Organizing the `wal_reader.c` file and specifically refactoring the WAL iteration logic.

Right now I'm organizing the `wal_reader.c` file. My current issue is that I can't run the `valgrind` tool on Mac to detect the memory issues. I'm trying to run the project inside a fedora docker image but I'm not sure how trustworthy the results of running in a docker image are.

**7/July/2024**

- Fixed the issue with the record header page separation, and a couple of consequential bugs. 
- Started refactoring, the refactoring process includes:

1- Move newly added `.h` files to a separate folder. 
2- Fix the naming of Structs to be consistent with the rest of the code. 
3- Fix struct definition styles.
4- Organize `.h` files organization in terms of the order of functions, macros, and flags.
5- Organize the `wal_reader.c` file and specifically refactor the WAL iteration logic.

Until now I'm refactoring the `.h` files and the progress is around 10 out of 24 files. 

**30/June/2024**
- This week I finished the implementation of `rm_desc` for all of the remaining resource managers, including:

1. Storage
2. CLOG
3. Database
4. Tablespace
5. MultiXact
6. RelMap
7. Hash
8. Gist
9. Sequence
10. SPGist
11. BRIN
12. CommitTs
13. ReplicationOrigin
14. Generic
15. LogicalMessage

- I was also able to generate a wall segment to include most of the resource managers and check if the code can parse them correctly. Currently, I'm dealing with a rare case where I guess the record's header is separated between two pages (instead of data, which was handled before).
- For the following week, after fixing these issues, I start refactoring the code. 

**23/June/2024**

- This week I was implementing the `rm_desc` function for different resource managers, until now, I have implemented it for the following list:

1. XLOG
16. Transaction
17. Heap
18. Heap2
19. Btree
20. Gin
21. Standby

- There are 22 resource managers in total, implementing it for the rest of them doesn't need a lot of creativity and it's mostly about trying to do it in a clean way to make the next step (refactoring) easier. 
- Did some minor refactors where needed for the newly added functions.

**16/June/2024**

1. Implemented the `rm_desc` for `Btree` records following `Standby`. 
22. Implement retrieving the data portion of records that are split between two pages and load them into the buffer. 
23. After implementing the first item, I noticed a bug in the parsing of the data portion of some `Heap` and `Btree` records, I spent most of my time this week figuring out the issue and fixing it, and was successful at the end :). The problem is that some of the flags such as `BKPIMAGE_IS_COMPRESSED` are used in Postgres 14 but are changed in the newer versions, I was parsing a record generated by Postgres 14 with the flags from the latest version in [here](https://doxygen.postgresql.org/xlogrecord_8h.html).
24. Implemented the description for the `BkpBlock` functionality and compared it to the actual `pg_waldump` and verified that it's working correctly. 
25. Pushed an up-to-date, squashed, and rebased version of the code into `parse-wal-records-squashed` branch [here](https://github.com/sh-soltanpour/pgmoneta/tree/parse-wal-records-squashed).

**09/June/2024**

1. This week was a good one in terms of my progress. I was able to fix the issues I had last week. For the offset calculation problem, I found that I needed to use the `MAXALIGN` function, and this function aligned the mentioned index from 90 to 96. Afterward, the next records were being parsed correctly. 
26. After this, I was able to iterate the WAL segment record by record and print each record's header, I was able to also skip the parts that a record was separated between two pages and handled these cases correctly. 
27. Then, I started decoding each WAL record, I mostly re-implemented the logic used in the [DecodeXLogRecord](https://doxygen.postgresql.org/xlogreader_8c.html#a940f02d24b9a445e2c57460e95451aab) function in a simplified and lighter version in Pgmoneta. 
28. I verified the correctness of the result by comparing the detected resource manager with the expected one by comparing my output against `pg_waldump`, I also compared and verified the record total length. 
29. Implemented an initial version of the `rm_desc` function for the `Standby` resource manager, I want to use this to make sure that the data is also parsed correctly, This was successful for the `Standby` resource manager, and for the next week, I'm going to implement it for other resource managers as well. Note that the code still needs a lot of cleaning, I'm planning to refactor it at some point before merging and after where we have the basic expected functionality. 

**02/June/2024**

1. Continuing the work from the previous week, I was able to parse the first `XLogRecord`. but I can't use the `xl_tot_len` field to jump over the first record and get the second `XLogRecord`. For instance, I'm at offset 40 of the first page and I read the first `XLogRecord`, and the `xl_tot_len` is 50, so I expect the next record to start at offset 90 = 40 + 50, but the parsed value in that offset is not correct, I also tried various ways such as 40 + 50 - 24 or 40 + 50 + 24 (24 is the header size), or even and removing 2 padding bytes, but none of them worked.
3. So I started reading the code of `XLogFindNextRecord`, `ReadInternalPage`, and other dependent functions in the `pg_waldump` source code. The difficult part of this process is figuring out what is useful in our case and what should be skipped, as well as navigating through a lot of nested functions. 
4. If someone knows about the first item and what's wrong with my calculations, or has useful resources on the second item that make it easier to understand the code, I would be more than happy if you could share them with me. 

**26/May/2024**

1. Follow [parse-wal-records](https://github.com/sh-soltanpour/pgmoneta/tree/parse-wal-records) for my most up-to-date progress in implementing the WAL parser.
3. Implemented the basic mechanism to go page by page and parse the header of each page. (`XLogLongPageHeaderData`, and `XLogPageHeaderData`). Verified the result by comparing the `xlp_magic` value against [the expected value](https://pgpedia.info/x/XLOG_PAGE_MAGIC.html).  
4. Started the implementation of the `XLogRecord` fixed header of each record, until now, I was able to parse the fixed header of the first record and figured out how to get the resource manager data from the parsed header.
5. I haven't committed the code of step 3 yet as I'm struggling with how to skip the rest of the first record and get the fixed header of the second one.  

**20/May/2024**
Find the longer version [here](https://github.com/pgmoneta/pgmoneta/discussions/288#discussioncomment-9500013), in summary:

1.  Learn more about how WAL segments and records are conducted and named. 
6.  Read through the `pg_waldump` [source code](https://doxygen.postgresql.org/pg__waldump_8c_source.html).
30. Try to implement a WAL segment parser from scratch, it was not really successful :)
31. Read through the [source code](https://doxygen.postgresql.org/pg__rewind_8c_source.html) of `pg_rewind` and watch [some videos](https://www.youtube.com/watch?v=qc7HTiJm_tQ) on how it can be used in production.
32. I talked to @jesperpedersen in the thread mentioned and decided to go with record extraction from WAL segments. Created a parent issue for the whole milestone #296 and broke it into three child issues #299, #300, and #301.

**17/May/2024**

1- I have been able to run Pgmoneta on my Mac (Kudos to @GuChad369 for this [PR](https://github.com/pgmoneta/pgmoneta/pull/237), it made my life easier a lot so I don't need to run a Fedora OS virtually).
2- As @jesperpedersen suggested, I have enabled compression and encryption to make sure these scenarios are covered. 
3- Have gone through the `wal.c` file in the project, to get some idea about how the WAL file is streamed.

For the next step, I want to parse the content of a WAL file streamed into Pgmoneta and load it to WAL records (issue: #296).

In general, I think I have a roadmap in my mind about what issues need to be created for the project to be done, but I need to organize my thoughts and ideas until May 20th, then I can create the list of issues here in GitHub so the mentors can comment on it and we may also be able to use the community help to do some of them.

**Comment by jesperpedersen** [comment]
Great ! 

Try and do a setup where both compression and encryption are enabled such that you can make sure that both scenarios are covered when reading the WAL files.

Welcome to GSoC 2024 !

**Comment by sh-soltanpour** [comment]
Hi @jesperpedersen, 
During the last two days, I've been working on how to read WAL files and parse them. I started with `pg_waldump`. At first, I felt the code was too complicated for our use case and decided to write my own parser from scratch. But after understanding the cases such as variable headers and data types and sizes, and continuous data over multiple pages, I feel that the code is getting more and more similar to what already exists in `pg_waldump` [here](https://doxygen.postgresql.org/pg__waldump_8c_source.html). 

I know that `pg_waldump` is using many shared resources from PostgreSQL and we need to re-implement them in our codebase if we aim to keep Pgmoneta independent. 

On the other hand,  I started reading the `pg_rewind` code as well. As far as I figured out, it tries to find a common ancestor between two databases using the information in the `pg_control` file and then tries to apply the source WAL files from the forked point to the target. If I understood correctly, `pg_rewind` does not need to parse all of the WAL segments to records and does its job on a segment level. 


Based on this, I see two areas that I should choose to put my focus on first:
1- If we want PiTR, we can proceed with something similar to what `pg_rewind` does for now. Then, we can implement segment parsing later when we want to extract logical information.
2- Adding the complete parsing ability to Pgmoneta so it will be able to parse WAL records from segments, and then go for the implementation of PiTR.

In general, I guess the first approach is more like a low-hanging fruit compared to the second one. 


Would be happy to know your opinion and corrections where I'm wrong, so we can set the goals and create the issues needed based on the chosen path. 

Thanks!

**Reply by jesperpedersen** [reply]
We need to understand the full WAL format without having dependencies on PostgreSQL code, and its data structures.

But, we need to do it in small steps and improve over time - maybe that means opening the file parse the header, skip the body and go from there.

As we discussed, this is a hard project - so we need to figure out how to best tackle it.

The end-goal is the per WAL record apply criteria based on the given input, so we need to know each WAL record

**Reply by sh-soltanpour** [reply]
Sounds good thank you!
So, I think I can break [this issue](https://github.com/pgmoneta/pgmoneta/issues/296) into these issues:
1- Implement extraction of WAL segment general portion headers
2- Implement extraction of the header part of the data portion
3- Implement extraction of the data part of data portion (This may need to be separated into two issues for block data and main data as well)

Please let me know if these issues make sense to you and if you have any suggestions, then I will add them. 
This should cover the first milestone of the project which is parsing and understanding WAL.

I'm thinking about creating the issues for the next milestone (which is iterating over the WAL records and selectively applying them) when we get there and have more info on the implementation. What do you think?

![image](https://github.com/pgmoneta/pgmoneta/assets/24779132/fcbab711-353b-4020-9bbe-ce24a6c3421b) [source](https://www.interdb.jp/pg/pgsql09/04.html)

**Reply by jesperpedersen** [reply]
Sounds like a good plan.

We can start to merge patches once there is something functional complete that we need for a foundation for a later patch.

I think it is very important for this work to have an active work branch registered in the discussion, or as a WIP pull request if close

**Comment by sh-soltanpour** [comment]
Hi @jesperpedersen,
I have implemented the codebase that is able to parse the page headers and verified the correctness using the `xlp_magic` value.
 I've pushed it to my forked repo in the `parse-wal-page-headers` branch. 

I wanted to create a main branch for the whole WAL parsing functionality, so finally it will be merged to `main`, but I don't have access to create a new branch on the original repo. So I think you can create the main working branch named `parse-wal-records` and I will make small patches and PRs to it?

**Reply by jesperpedersen** [reply]
The important is to have documented work branches for your work.

They will be in your repository, but documented in this discussion.

Keep your work off `main` since that will be updated frequently, and you want to rebase against it as often as possible.

Breaking down the work into smaller pieces is always good - we can merge functionality as soon as it is complete and make sense to be included in the official release.

I'm expecting that the initial patch will be fairly large, and that is why a documented work branch is important to have. If you go too much down the wrong path, then it will take a lot of time to change...

You can update your work log when you even pushed something like an important bug fix

**Comment by jesperpedersen** [comment]
I think you are making good progress !

Remember that you have to support the version range so,

* PostgreSQL 13: 0xD106
* PostgreSQL 14: 0xD10D
* PostgreSQL 15: 0xD110
* PostgreSQL 16: 0xD113
* PostgreSQL 17: 0xD115

Also, it would be good to have a work branch where your commits are squashed, and rebased against main - this will make it easier to test. See https://github.com/pgmoneta/pgmoneta/blob/main/doc/DEVELOPERS.md

Use license headers, and uncrustify to get your coding style closer to the official one.

**Reply by sh-soltanpour** [reply]
I have pushed a squashed version of the code [here](https://github.com/sh-soltanpour/pgmoneta/tree/parse-wal-records-squashed), will keep it rebased with the main branch and merge and squash changes into it weekly. Also applied uncrustify and license headers.

For this week, I'm going to focus on implementing `rm_desc` for different resource managers and debug them if needed, to make sure the data is parsed correctly, 

Then I start refactoring the code, testing it with different versions, and updating the incompatible parts to work based on the Postgres version.

**Reply by jesperpedersen** [reply]
Sounds like a good plan !

**Reply by jesperpedersen** [reply]
@GuChad369 is working a test suite setup that will help us - follow @ https://github.com/pgmoneta/pgmoneta_ext/pull/11

**Comment by jesperpedersen** [comment]
@sh-soltanpour How is it going ?

**Reply by sh-soltanpour** [reply]
Hey Jesper!
just updated the progress report for this week. 

In summary, worked on adding the rm_desc function for different resource managers. There are 22 in total, so I’ve still got some to go, the process is kinda straightforward but time-consuming as there are lots of resource managers, and just need to make sure to keep things clean to make the upcoming refactoring easier.

**Comment by jesperpedersen** [comment]
@sh-soltanpour 

When you start to refactor I think it would be best to keep it in one pair - e.g. one `.h` and one `.c` - and use comments on where they originally came from.

Since this is a VERY complex area the API should be as simple as possible, so we will likely have a lot of iterations over that.

Remember to squash, and rebase as you go along...

**Reply by sh-soltanpour** [reply]
Hi @jesperpedersen ,

Thanks for the comment. I'm not 100% sure how we can reduce the API complexity by putting all of them in one file, since the functions defined in the single `.h` file are exposed anyway. But I agree that we can keep them separated like in a new directory. What do you think?

**Reply by jesperpedersen** [reply]
We can start with an `include/wal/` directory for your `.h`'s and go from there.

I think the main API (under `include/`) should be some sort of streaming API, so `open`, `iterate` and `close` where each WAL record can be identified using the main API as well, and can be type cast'ed into its proper type if we need that.

We can always to look at your squashed branch in more wehn you are happy with it

**Comment by jesperpedersen** [comment]
Two minor comments.

We don't use

```
typedef struct X
{
} x;
```

just

```
struct X
{
};
```

And we don't use CamelCase - I know that PostgreSQL/Core does that, but I think we should keep our style.

**Comment by jesperpedersen** [comment]
We have to start to think about `write()` too...

If we have a common "header" aka base `struct` that just have key properties - like the type - maybe we can use a `sizeof(struct X)` and delegate to a function that write out the raw byte array from that...

I think that the first test case would be to read a WAL segment, and write it out again, then byte compare the two

**Reply by sh-soltanpour** [reply]
Yeah, this is a good next step, I guess it won't be such hard work as we already have the data structs and need to convert them to bytes and place them correctly in pages!

**Comment by jesperpedersen** [comment]
Can you move the `.c` files to `libpgmoneta/wal/` ?

**Reply by sh-soltanpour** [reply]
Yep, will do

**Comment by jesperpedersen** [comment]
Remember to rebase such you have access to the test suite for your stuff

**Reply by sh-soltanpour** [reply]
done, thanks!

**Comment by jesperpedersen** [comment]
I think you can trust `valgrind` from within the container...

**Comment by jesperpedersen** [comment]
> Regarding what's remaining, I can say that there are some defined values in my code that should be parsed from the pg_control file I guess, and I've used constant values for them at the moment, and also the code only supports Postgres 14 right now and the compatibility should get extended to the 13-17 versions. With these, do you think that we can start having a PR and reviewing it or do you prefer these to be added and then check them all at once?

We know the version from https://github.com/pgmoneta/pgmoneta/blob/main/src/include/pgmoneta.h#L218 and I think it would good to support 13 to 17 in the initial pull request.

Then we have the WAL records, and then we can look at the "top-level" API in more detail.

I think you have most in place, so now it is more about making it clear what is different between the different WAL versions

**Comment by jesperpedersen** [comment]
@sh-soltanpour We likely need infrastructure around `pg_control` to parse it into a pgmoneta data type, but we can start by bringing in the 13-17 WAL records, and work from there.

What do you think is best ?

**Reply by sh-soltanpour** [reply]
Yeah, I agree that we can start merging the 13-17 WAL records, and then create some issues to parse `pg_control` data.

**Reply by jesperpedersen** [reply]
If you think everything can be committed under #296, then you can just close #299, #300 and #301

**Reply by jesperpedersen** [reply]
pgmoneta 0.13 is out, so we can start to bring this in

**Reply by jesperpedersen** [reply]
Remember to squash and rebase before submitting the pull request
